packages/jubmoji-api/src/proofs.ts

```diff
@@ -0,0 +1,313 @@
+      usedSigNullifiers: usedSigNullifiers?.map(hexToBigInt),
+      pathToCircuits: this.pathToCircuits,
+    });
+  }
+}
+
+export class NUniqueJubmojisInCollection {
+  collectionPubKeys: string[];
+  sigNullifierRandomness: string;
+  N: number;
+  pathToCircuits?: string;
+
+  constructor({
+    collectionPubKeys,
+    sigNullifierRandomness,
+    N,
+    pathToCircuits,
+  }: NUniqueJubmojiInCollectionClassArgs) {
+    this.collectionPubKeys = collectionPubKeys;
+    this.sigNullifierRandomness = sigNullifierRandomness;
+    this.N = N;
+    this.pathToCircuits = pathToCircuits;
+  }
+
+  async prove({
+    jubmojis,
+  }: NUniqueJubmojiInCollectionProofArgs): Promise<NUniqueJubmojiInCollectionProof> {
+    const pubKeyNullifierRandomness = getRandomNullifierRandomness();
+
+    const sigs = [];
+    const msgHashes = [];
+    const publicInputs = [];
+    const indices = [];
+    for (const jubmoji of jubmojis) {
+      const { sig, msgHash, pubKey, R, T, U } =
+        getMembershipProofArgsFromJubmoji(jubmoji);
+      sigs.push(sig);
+      msgHashes.push(msgHash);
+      publicInputs.push({ R, T, U });
+      indices.push(this.collectionPubKeys.indexOf(pubKey));
+    }
+
+    const merkleProofs = getMerkleProofListFromCache(
+      this.collectionPubKeys,
+      indices
+    );
+
+    const membershipProofs = await batchProveMembership({
+      sigs,
+      msgHashes,
+      publicInputs,
+      merkleProofs,
+      sigNullifierRandomness: hexToBigInt(this.sigNullifierRandomness),
+      pubKeyNullifierRandomness,
+      pathToCircuits: this.pathToCircuits,
+    });
+
+    return {
+      serializedMembershipProofs: membershipProofs.map(
+        serializeMembershipProof
+      ),
+    };
+  }
+
+  // Checks that all the membershipProofs are valid, and that their pubKeyNullifiers are unique
+  // while having the same pubKeyNullifierRandomnessHash
+  async verify({
+    serializedMembershipProofs,
+    usedSigNullifiers,
+  }: NUniqueJubmojiInCollectionProof): Promise<VerificationResult> {
+    // Must have at least N proofs
+    if (serializedMembershipProofs.length < this.N) {
+      return { verified: false };
+    }
+
+    const membershipProofs = serializedMembershipProofs.map(
+      deserializeMembershipProof
+    );
+
+    // Checks that all the pubKeyNullifiers are unique but all the
+    // pubKeyNullifierRandomnessHashes are the same
+    const pubKeyNullifiers = membershipProofs.map(
+      (proof) => getPublicSignalsFromMembershipZKP(proof.zkp).pubKeyNullifier
+    );
+    if (!areAllBigIntsDifferent(pubKeyNullifiers)) {
+      return { verified: false };
+    }
+
+    const pubKeyNullifierRandomnessHashes = membershipProofs.map(
+      (proof) =>
+        getPublicSignalsFromMembershipZKP(proof.zkp)
+          .pubKeyNullifierRandomnessHash
+    );
+    if (!areAllBigIntsTheSame(pubKeyNullifierRandomnessHashes)) {
+      return { verified: false };
+    }
```

Comment:
This logic is so sick lol. We are gonna have such a banger blog post after Devconnect about all the circuit tricks